NitroxServer/GameLogic/Spawning/EntitySpawner.cs

```diff
@@ -6,123 +6,132 @@
-                entitySpawnPoints = BatchCellsParser.GetEntitySpawnPoints();
+                // TODO: If data files can't be loaded the code will crash due to NRE's.
+                batchCellsParser = new BatchCellsParser();
 
                 LootDistributionsParser lootDistributionsParser = new LootDistributionsParser();
                 lootDistributionData = lootDistributionsParser.GetLootDistributionData(lootDistributionString);
-
-                SpawnEntities();
             }
         }
 
-        public Dictionary<AbsoluteEntityCell, List<Entity>> GetEntitiesByAbsoluteCell()
-        {
-            return entitiesByAbsoluteCell;
-        }
-
-        private void SpawnEntities()
+        /// <summary>
+        /// Spawns entities for <paramref name="batchId"/> if it was not loaded yet.
+        /// </summary>
+        public Optional<IDictionary<AbsoluteEntityCell, List<Entity>>> SpawnUnloadedEntitiesForBatch(Int3 batchId)
         {
-            Log.Info("Spawning entities...");
-            entitiesByAbsoluteCell = new Dictionary<AbsoluteEntityCell, List<Entity>>();
-            Random random = new Random();
-
-            foreach (EntitySpawnPoint entitySpawnPoint in entitySpawnPoints)
+            lock (parsedBatches)
             {
-                DstData dstData;
-                if (!lootDistributionData.GetBiomeLoot(entitySpawnPoint.BiomeType, out dstData))
+                if (parsedBatches.Contains(batchId))
                 {
-                    continue;
+                    return Optional<IDictionary<AbsoluteEntityCell, List<Entity>>>.Empty();
                 }
+                parsedBatches.Add(batchId);
+            }
 
-                float rollingProbabilityDensity = 0;
+            Log.Debug("Batch {0} not parsed yet; parsing...", batchId);
 
-                PrefabData selectedPrefab = null;
+            IDictionary<AbsoluteEntityCell, List<Entity>> entitiesByAbsoluteCell = new Dictionary<AbsoluteEntityCell, List<Entity>>();
 
-                foreach (PrefabData prefab in dstData.prefabs)
+            foreach (EntitySpawnPoint esp in batchCellsParser.ParseBatchData(batchId))
+            {
+                List<Entity> entities;
+                if (!entitiesByAbsoluteCell.TryGetValue(esp.AbsoluteEntityCell, out entities))
                 {
-                    float probabilityDensity = prefab.probability / entitySpawnPoint.Density;
-                    rollingProbabilityDensity += probabilityDensity;
+                    entitiesByAbsoluteCell[esp.AbsoluteEntityCell] = entities = new List<Entity>();
                 }
 
-                double randomNumber = random.NextDouble();
-                double rollingProbability = 0;
+                entities.AddRange(SpawnEntities(esp));
+            }
+
+            return Optional<IDictionary<AbsoluteEntityCell, List<Entity>>>.Of(entitiesByAbsoluteCell);
+        }
+
+        private IEnumerable<Entity> SpawnEntities(EntitySpawnPoint entitySpawnPoint)
```

Comment:
minor: For my understanding (As I am not a c# programmer), why do we want to use IEnumerable with yield operator over returning a collection?  From my understanding, yield operator implies to the reader that we want the code to have deferred execution.  However, in this usage we are simply doing AddRange which takes all of the elements anyways.

Reply:
It's not something C# specific, and `yield` has many more usecases than deferred execution. I'm not fond of creating and returning lists because it's too verbose, and wasting time/memory because it's only used for a very short amount of time; once the function returns the items are copied to the other collection after which the list will be GC'd. An enumerable solves that nicely, as the items are evaluated and and directly stored in the right collection. Yield is a perfect, concise way to get such an enumerable. And what would you do in case of the `yield break` at the beginning of the function? Return `null`? Return an empty list?

In any case, I wasn't expecting a merge already, and still wanted to convert this to a oneliner:

```cs
batchCellsParser
    .ParseBatchData(batchId)
    ‎.GroupBy(esp => esp.AbsoluteEntityCell)
    ‎.ToDictionary(grp => grp.Key, grp.SelectMany(SpawnEntities).ToList());
```
(From the top of my head)